rfcs/003-double-entry-ledger.md

```diff
@@ -0,0 +1,149 @@
+
+This change will increase accuracy because it will simplify the way we
+store the transactions. It will also decrease duplicated code (likely
+reducing bugs) because it will provide a library that implements all
+the basic operations related to the ledger.
+
+*Fig 1.0 - Payment Flow*
+![Payment Flow](./imgs/payment-flow.svg "Payment Flow")
+
+## Goals
+
+1. Increase accuracy of the ledger
+   1. Create collective type "Payment Provider"
+   2. Create separate transactions for fees
+2. Make it easier to build new features that depend on the ledger
+   1. `libocledger`: wrap the most common operations related to the
+      ledger under a library that can be used across all the places
+      that interact with the ledger and decrease duplicated code in
+      the repository.
+
+### Create collective type "Payment Provider"
+
+As mentioned before, the amount of the payment processor fee is stored
+in the field `paymentProcessorFeeInHostCurrency` in the `Transactions`
+table. To find out to which specific payment method the fee went to,
+the `Transactions` table also records the `PaymentMethodId`.
+
+This is different from how transactions are tracked between User &
+Collective. To address this difference, the Payment Provider needs to
+have a ledger as well.
+
+### Create separate transactions for fees
+
+#### Use case example (Order)
+
+Given a transaction that represents moving $50 dollars from User to
+Collective with 5% of platform fee, 10% of host fee and 2.9%+30c of
+payment processor fee.
+
+##### How it is currently stored
+
+|            |  User | Collective |
+|------------|------:|-----------:|
+| Type       | DEBIT |     CREDIT |
+| Amount     | -4075 |       5000 |
+| Host Fee   |  -500 |       -500 |
+| Plat Fee   |  -250 |       -250 |
+| PP Fee     |  -175 |       -175 |
+| Net Amount | -5000 |       4075 |
+
+##### How it should look like after this change
+
+|  User | Payment Provider | Platform |  Host | Collective |
+|------:|-----------------:|---------:|------:|-----------:|
+| -5000 |             5000 |          |       |            |
+|       |             -250 |      250 |       |            |
+|       |            -4575 |          |  4575 |            |
+|       |                  |          | -4325 |       4325 |
+
+#### Use case example (Expense)
+
+Given a transaction that represents moving $50 dollars from Collective
+to User with 2.9%+30c of payment processor fee.
+
+##### How it is currently stored
+
+|            | Collective |   User |
+|------------|-----------:|-------:|
+| Type       |      DEBIT | CREDIT |
+| Amount     |      -5000 |   5000 |
+| PP Fee     |       -175 |   -175 |
+| Net Amount |      -5175 |   5000 |
+
+##### How it should look like after this change
+
+| Collective |  Host | Payment Provider | User |
+|-----------:|------:|-----------------:|-----:|
+|      -5000 |  5000 |                  |      |
+|            | -5175 |             5175 |      |
```

Comment:
FYI, the payment provider fees comes from the collective, not the host.

Reply:
Oh Thanks! This brings me to a question though. If each row in that table represents a Transaction row, what would you think that is more appropriate?

### To create an explicit transaction for the fee?
```
| Collective |  Host | Payment Provider | User |
|-----------:|------:|-----------------:|-----:|
|      -5000 |  5000 |                  |      |
|            | -5000 |             5000 |      |
|       -175 |       |              175 |      |
|            |       |            -5000 | 5000 |
```

### To assume the fee is whatever didn't get sent to the user:
```
| Collective |  Host | Payment Provider | User |
|-----------:|------:|-----------------:|-----:|
|      -5175 |  5175 |                  |      |
|            | -5175 |             5175 |      |
|            |       |            -5000 | 5000 |
```